### PR TITLE
website: update metrics doc

### PR DESCRIPTION
**Pull request by lucarin91**
#### :question: What

Update metrics documentation

### CODE REVIEW

File: content/en/docs/How Tos/metrics.md [hunk 1]
```diff
@@ -134,13 +73,161 @@ or an error saying why it can't.
...
+Baker supports two main ways to expose custom metrics, namely:
+
+- return a 
+[`baker.MetricsBag`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsBag)
+instance from the `Stats` method 
+- directly use the 
+[`baker.MetricsClient`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsClient) in 
+the component code
+
+The two mechanisms following pull vs push approach respectively.
```
Comment (tommyblue) [inline content/en/docs/How Tos/metrics.md:113]:
```suggestion
The two mechanisms follow the pull vs push approach respectively.
```

File: content/en/docs/How Tos/metrics.md [hunk 2]
```diff
@@ -134,13 +73,161 @@ or an error saying why it can't.
...
+The `MetricsBag` should be returned by the `Stats` method along with the default
+statistics and it will be collected 1 per second by the Baker runtime. Differently,
```
Comment (tommyblue) [inline content/en/docs/How Tos/metrics.md:115]:
```suggestion
statistics and it will be collected once per second by the Baker runtime. Differently,
```

File: content/en/docs/How Tos/metrics.md [hunk 3]
```diff
@@ -134,13 +73,161 @@ or an error saying why it can't.
...
+the `MetricsClient` can be requested from the Baker topology in the component
+instantiation and it can be used in any part of the component code to report metrics.
+If there is no particular requirements it is suggested to prefer the `MetricsBag`.
```
Comment (tommyblue) [inline content/en/docs/How Tos/metrics.md:118]:
```suggestion
If there is no particular requirement it is suggested to prefer the `MetricsBag` approach. 
```

or `are -> requirements`

File: content/en/docs/How Tos/metrics.md [hunk 4]
```diff
@@ -134,13 +73,161 @@ or an error saying why it can't.
...
+[`baker.MetricsClient`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsClient) in 
+the component code
+
+The two mechanisms following pull vs push approach respectively.
+The `MetricsBag` should be returned by the `Stats` method along with the default
+statistics and it will be collected 1 per second by the Baker runtime. Differently,
+the `MetricsClient` can be requested from the Baker topology in the component
+instantiation and it can be used in any part of the component code to report metrics.
+If there is no particular requirements it is suggested to prefer the `MetricsBag`. 
+Indeed, the pull approach permits the reduction the metrics overhead during the
```
Comment (tommyblue) [inline content/en/docs/How Tos/metrics.md:119]:
```suggestion
Indeed, the pull approach permits the reduction of the metrics overhead during the 
```

File: content/en/docs/How Tos/metrics.md [hunk 5]
```diff
@@ -134,13 +73,161 @@ or an error saying why it can't.
...
+The two mechanisms following pull vs push approach respectively.
+The `MetricsBag` should be returned by the `Stats` method along with the default
+statistics and it will be collected 1 per second by the Baker runtime. Differently,
+the `MetricsClient` can be requested from the Baker topology in the component
+instantiation and it can be used in any part of the component code to report metrics.
+If there is no particular requirements it is suggested to prefer the `MetricsBag`. 
+Indeed, the pull approach permits the reduction the metrics overhead during the 
+record processing.
+
+Both `MetricsBag` and `MetricsClient` Metrics supports the most common metric types,
```
Comment (tommyblue) [inline content/en/docs/How Tos/metrics.md:122]:
```suggestion
Both `MetricsBag` and `MetricsClient` Metrics support the most common metric types,
```

File: content/en/docs/How Tos/metrics.md [hunk 6]
```diff
@@ -134,13 +73,161 @@ or an error saying why it can't.
...
+namely:
+- `RawCounter`, a cumulative counter that can only increase.
+- `DeltaCounter`, the total number of event occurrences in a unit time.
+- `Gauge`, a snapshot of an event in the last unit time. 
+- `Histogram`, a statistical distribution of a set of values in one unit 
+of time.
+- `Duration` or `Timing`, like an histogram but with time durations
```
Comment (tommyblue) [inline content/en/docs/How Tos/metrics.md:129]: i'd replace the comma with colon and remove the trailing dot

File: content/en/docs/How Tos/metrics.md [hunk 7]
```diff
@@ -134,13 +73,161 @@ or an error saying why it can't.
...
+
+Moreover, the `MetricsClient` is the direct interface to the remote metrics service 
+used by the topology, e.g., *Datadog*, thus it supports some specific features such 
+as **tags**.
+Tags are a way of adding dimensions to telemetries so they can be filtered, 
+aggregated, and compared in different visualizations.
+Therefore, if a component requires to publish its metrics with a set of specific tags, the 
+`MetricsClient` should use rather than `MetricsBag`.
```
Comment (tommyblue) [inline content/en/docs/How Tos/metrics.md:137]:
```suggestion
`MetricsClient` should be used rather than `MetricsBag`.
```

File: content/en/docs/How Tos/metrics.md [hunk 8]
```diff
@@ -134,13 +73,161 @@ or an error saying why it can't.
...
+However, it is suggested to use `MetricsClient` in the `Stats` method of the 
+components to reduce contention and improve performance in the processing of the 
+elements.
```
Comment (tommyblue) [inline content/en/docs/How Tos/metrics.md:140]: together with the above sentence, the whole concept seems a bit fuzzy, I'd maybe rephrase a bit

File: content/en/docs/How Tos/metrics.md [hunk 9]
```diff
@@ -134,13 +73,161 @@ or an error saying why it can't.
...
+there are some situations in which it is preferable the `MetricsClient`, namely:
+- the components need to publish the metrics using specific tags,
+- the components cannot centralize the metric collections, e.g. the component has 
+multiple goroutines.
+
+#### MetricsBag Example
+
+Let's say our filter needs to perform HTTP requests in order to decide whether a 
+record should be discarded, we might want to keep track of the requests' durations 
+in a histogram. In this case, we would probably record a slice of `time.Duration` in
```
Comment (tommyblue) [inline content/en/docs/How Tos/metrics.md:152]:
```suggestion
in an histogram. In this case, we would probably record a slice of `time.Duration` in 
```

File: content/en/docs/How Tos/metrics.md [hunk 10]
```diff
@@ -134,13 +73,161 @@ or an error saying why it can't.
...
+Let's say our filter needs to perform HTTP requests in order to decide whether a 
+record should be discarded, we might want to keep track of the requests' durations 
+in a histogram. In this case, we would probably record a slice of `time.Duration` in 
+our filter and call 
+[`AddTimings`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsBag.AddTimings) on
+the returned `MetricsBag`.
+
+An important point is that Baker may call `Process` and `Stats` concurrently, from 
+different goroutines so you must use proper locking on data structures that are
+shared between these two methods.
```
Comment (tommyblue) [inline content/en/docs/How Tos/metrics.md:159]: or use atomic values. Maybe I'd mention it

File: content/en/docs/How Tos/metrics.md [hunk 11]
```diff
@@ -134,13 +73,161 @@ or an error saying why it can't.
...
+instance has been successfully created, it's made available to and used by a Baker 
+pipeline to report metrics. During construction, components receive the 
+[`MetricsClient`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsClient) instance.
 [`baker.MetricsClient`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsClient) 
-supports the most common type of metric types: *gauges*, *counters* and *histograms*.
+provides a set of methods to report the most common type of metric types (e.g. 
+*gauges*, *counters* and *histograms*)
+
+Taking the previous example of the `MetricsBag`, we now consider the situation in 
+which the time duration that we want to collect, needs to be repported along with specific
```
Comment (tommyblue) [inline content/en/docs/How Tos/metrics.md:205]:
```suggestion
which the time duration that we want to collect needs to be repported along with specific 
```

**Review by tommyblue — APPROVED** [review]
good reworking, much clearer now. I added some language fixes/tips.
I'd also perform a Grammarly check to spot more errors

File: content/en/docs/How Tos/metrics.md [hunk 12]
```diff
@@ -134,13 +73,160 @@ or an error saying why it can't.
...
+aggregated, and compared in different visualizations.
+Therefore, if a component requires to publish its metrics with a set of specific tags, the 
+`MetricsClient` should be used rather than `MetricsBag`.
+However, it is suggested to use the `MetricsClient` in the `Stats` method to limit
+the number of calls to the client. Indeed, calling the `MetricsClient` methods inside the 
+`Process` method could introduce overheads and performance degradations.
+
+In summary, the go-to way is to implements custom statistics with a `MetricsBag`, but 
+there are some situations in which it is preferable the `MetricsClient`, namely:
+- the components need to publish the metrics using specific tags,
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:143]:
```suggestion
- your component needs to publish the metrics using a set of tag that changes at runtime
```

File: content/en/docs/How Tos/metrics.md [hunk 13]
```diff
@@ -134,13 +73,160 @@ or an error saying why it can't.
...
+- the components cannot centralize the metric collections, e.g. the component has
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:144]:
```suggestion
- you can't centralize metrics collection in the `Stats` method since the metrics to expose are produced by different worker goroutines running inside your component 
```

File: content/en/docs/How Tos/metrics.md [hunk 14]
```diff
@@ -134,13 +73,160 @@ or an error saying why it can't.
...
+Let's say our filter needs to perform HTTP requests in order to decide whether a 
+record should be discarded, we might want to keep track of the requests' durations 
+in an histogram. In this case, we would probably record a slice of `time.Duration` in 
+our filter and call 
+[`AddTimings`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsBag.AddTimings) on
+the returned `MetricsBag`.
+
+An important point is that Baker may call `Process` and `Stats` concurrently, from 
+different goroutines so you must use atomics or proper locking on data that are 
+shared between these two methods.
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:158]:
```suggestion
An important point is that Baker may call `Process` and `Stats` from different goroutines
so access to the shared data must be properly synchronized (atomic, locks, channel, etc.)
```

**Review by arl — COMMENTED** [review]
Good Job.
I let many comments most of which are suggestions

**Comment by lucarin91** [comment]
I did my best to implement all the suggestions, but I end up rewriting part of the text, feel free to point out somethings that I miss.

File: content/en/docs/How Tos/metrics.md [hunk 15]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+Due to historical reasons, these fields have the word _lines_ in them but they do 
+mean the number of records.
+
+The code example of 
+[how-to-create-filter](https://getbaker.io/docs/how-tos/create_filter/#processing-records) 
+shows how to correctly report metrics in a custom Filter.
+
+### Report custom metrics
+
+In addition to the records counter described above, components can report custom metrics
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:102]:
```suggestion
In addition to the record counters described above, components can report custom metrics
```

File: content/en/docs/How Tos/metrics.md [hunk 16]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+giving a more specific view about the component health or performance.
+Baker supports two ways to expose custom metrics, namely:
+
+- return a 
+[`baker.MetricsBag`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsBag)
+instance from the `Stats` method 
+- directly use the 
+[`baker.MetricsClient`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsClient) in 
+the component code
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:111]: Slightly better wording, plus bulleted lists items must be punctuated as normal sentences: 
```suggestion
Baker supports two ways of exposing custom metrics:

- returning a 
[`baker.MetricsBag`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsBag)
instance from the `Stats` method.
- directly using the 
[`baker.MetricsClient`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsClient) in 
the component code.
```

File: content/en/docs/How Tos/metrics.md [hunk 17]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+
+The two mechanisms follow the pull vs push approach, respectively.
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:113]: I guess we should use plural here since we're not speaking about a single "push-pull" approach, but 2 different approaches: a push approach and a pull one, so:
```suggestion
The two mechanisms follow the pull vs push approaches, respectively.
```

File: content/en/docs/How Tos/metrics.md [hunk 18]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+so as to use it at any point in your component code since well-behaved `MetricsClient` 
+implementations are safe for concurrent use by multiple goroutines.
+
+Both `MetricsBag` and `MetricsClient` Metrics support the most common metric types,
+namely:
+- `RawCounter`: a cumulative counter that can only increase
+- `DeltaCounter`: the total number of event occurrences in a unit time
+- `Gauge`: a snapshot of an event in the last unit time
+- `Histogram`: statistical distribution of a set of values in one unit of time
+- `Duration` or `Timing`: like a histogram but with time durations
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:129]: dots after bulleted lists

File: content/en/docs/How Tos/metrics.md [hunk 19]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+
+If there is no particular requirement the `MetricsBag` approach is preferred. Indeed, 
+the pull approach of the `MetricsBag` is simpler and it better integrates with the
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:132]:
```suggestion
the `MetricsBag` pull approach is simpler and it better integrates with the 
```

File: content/en/docs/How Tos/metrics.md [hunk 20]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+- `RawCounter`: a cumulative counter that can only increase
+- `DeltaCounter`: the total number of event occurrences in a unit time
+- `Gauge`: a snapshot of an event in the last unit time
+- `Histogram`: statistical distribution of a set of values in one unit of time
+- `Duration` or `Timing`: like a histogram but with time durations
+
+If there is no particular requirement the `MetricsBag` approach is preferred. Indeed, 
+the pull approach of the `MetricsBag` is simpler and it better integrates with the 
+other default metric gathering. 
+In addition to that, the `MetricsClient` requires special attention to avoids too
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:134]:
```suggestion
In addition to that, the `MetricsClient` requires special attention to avoid too
```

File: content/en/docs/How Tos/metrics.md [hunk 21]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+many calls to the client during the record processing.
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:135]:
```suggestion
many calls to the client during record processing.
```

File: content/en/docs/How Tos/metrics.md [hunk 22]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+Indeed, the `MetricsClient` is a low-level and direct way to communicate with your 
+metrics system, thus it permits extra flexibility, but it skips the additional 
+aggregation/buffering layer of the `MetricsBag`.
+
+Moreover, the `MetricsClient` supports also the runtime tagging of the published metrics.
+Tags (or labels) are additional dimensions on metrics so they can be filtered, 
+aggregated, and compared in different visualizations.
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:142]:
```suggestion
Tags (or labels) are additional dimensions on metrics and allow them to be filtered, 
aggregated, and compared in different visualizations.
```

File: content/en/docs/How Tos/metrics.md [hunk 23]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+Therefore, if a component requires publishing its metrics with a set of specific tags,
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:143]: maybe _dynamic_ here explains the point better?
```suggestion
Therefore, if tags are only known at runtime and you need to add dynamic tags to your metrics, 
```

File: content/en/docs/How Tos/metrics.md [hunk 24]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+aggregation/buffering layer of the `MetricsBag`.
+
+Moreover, the `MetricsClient` supports also the runtime tagging of the published metrics.
+Tags (or labels) are additional dimensions on metrics so they can be filtered, 
+aggregated, and compared in different visualizations.
+Therefore, if a component requires publishing its metrics with a set of specific tags, 
+the `MetricsClient` should be used rather than `MetricsBag` (see the 
+[MetricsClient example](/docs/how-tos/metrics/#metricsclient-example)).
+
+In summary, the go-to way is to implements custom statistics with a `MetricsBag`, but
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:147]:
```suggestion
In summary, the go-to way is to implement custom statistics with a `MetricsBag`, but 
```

File: content/en/docs/How Tos/metrics.md [hunk 25]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+
+Moreover, the `MetricsClient` supports also the runtime tagging of the published metrics.
+Tags (or labels) are additional dimensions on metrics so they can be filtered, 
+aggregated, and compared in different visualizations.
+Therefore, if a component requires publishing its metrics with a set of specific tags, 
+the `MetricsClient` should be used rather than `MetricsBag` (see the 
+[MetricsClient example](/docs/how-tos/metrics/#metricsclient-example)).
+
+In summary, the go-to way is to implements custom statistics with a `MetricsBag`, but 
+there are some situations in which it is preferable the `MetricsClient`, namely:
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:148]:
```suggestion
there are some situations where a `MetricsClient` is preferred, for example:
```

File: content/en/docs/How Tos/metrics.md [hunk 26]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+- your component needs to publish the metrics using a set of tag that changes at runtime
+- you can't centralize metrics collection in the `Stats` method since the metrics 
+to expose are produced by different worker goroutines running inside your component 
+multiple goroutines.
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:152]: missing dots

File: content/en/docs/How Tos/metrics.md [hunk 27]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+
+Let's say our filter needs to perform HTTP requests in order to decide whether a 
+record should be discarded, we might want to keep track of the requests' durations 
+in an histogram. In this case, we would probably record a slice of `time.Duration` in 
+our filter and call 
+[`AddTimings`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsBag.AddTimings) on
+the returned `MetricsBag`.
+
+An important point is that Baker may call `Process` and `Stats` from different goroutines
+so access to the shared data must be properly synchronized (atomic, locks, channel, etc.)
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:164]:
```suggestion
so access to the shared data must be properly synchronized (atomic, lock, channel, etc.)
```

File: content/en/docs/How Tos/metrics.md [hunk 28]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
 
-Once a [`baker.MetricsClient`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsClient)
-instance has been successfully created, it's made available to and used by
-a Baker pipeline to report metrics. During construction, components receive the 
-[`MetricsClient`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsClient) instance.
+```go
+func (f *MyFilter) Process(r Record, next func(Record)) {
+    atomic.AddInt64(&myFilter.totalLines, 1)
+
+    /* perform http request */
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:170]:
```suggestion
    // Perform http request
```

File: content/en/docs/How Tos/metrics.md [hunk 29]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+    f.mu.Lock() // keep track of its duration
+    f.requestDurations = append(f.requestDurations, duration)
+    f.mu.Unlock()
+
+    if (/* filter logic*/) {
+        // discard line
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:176]:
```suggestion
        // Discard record
```

File: content/en/docs/How Tos/metrics.md [hunk 30]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+        atomic.AddInt64(&f.filteredLines, 1)
+        return
+    }
+}
+
+func (f *MyFilter) Stats() baker.FilterStats {
+    // copy the slice of durations
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:183]:
```suggestion
    // Copy the slice of durations
```

File: content/en/docs/How Tos/metrics.md [hunk 31]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+    }
+}
+```
+
+#### MetricsClient Example
 
+Once a 
 [`baker.MetricsClient`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsClient) 
-supports the most common type of metric types: *gauges*, *counters* and *histograms*.
+instance has been successfully created, it's made available to and used by a Baker
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:202]:
```suggestion
instance has been successfully created, it's made available to and can be used by a Baker 
```

File: content/en/docs/How Tos/metrics.md [hunk 32]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
 
+Once a 
 [`baker.MetricsClient`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsClient) 
-supports the most common type of metric types: *gauges*, *counters* and *histograms*.
+instance has been successfully created, it's made available to and used by a Baker 
+pipeline to report metrics. During construction, components receive the 
+[`MetricsClient`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsClient) instance.
+[`baker.MetricsClient`](https://pkg.go.dev/github.com/AdRoll/baker#MetricsClient) 
+provides a set of methods to report the most common type of metric types (e.g. 
+*gauges*, *counters* and *histograms*)
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:207]: the list is finite, so we can directly use:
```suggestion
provides a set of methods to report the most common type of metric types (*gauges*,
*counters* and *histograms*)
```

File: content/en/docs/How Tos/metrics.md [hunk 33]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+
+Taking the previous example of the `MetricsBag`, we now consider the situation in 
+which the time duration that we want to collect needs to be repported along with specific 
+tags.
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:211]:
```suggestion
Let's now consider we want to publish additional tags alongside the duration slice of our previous `MetricsBag` example.
```

File: content/en/docs/How Tos/metrics.md [hunk 34]
```diff
@@ -134,13 +72,171 @@ or an error saying why it can't.
...
+    return baker.FilterStats{
+        NumProcessedLines: atomic.LoadInt64(&f.totalLines),
+        NumFilteredLines: atomic.LoadInt64(&f.filteredLines),
+    }
+}
+```
 
+In the example, we prefered to use the `MetricsClient` in the `Stats` method to limit
+the number of calls to the client. Indeed, calling the `MetricsClient` methods inside 
+the `Process` method could introduce overheads and performance degradations.
```
Comment (arl) [inline content/en/docs/How Tos/metrics.md:242]:
```suggestion
Here, we chose to use the `MetricsClient` in the `Stats` method to limit the number of
calls to the client. Indeed, since `Stats` gets called once per second, the `MetricsClient` is 
also called once per second. We could have directly called `DurationWithTags` from the 
`Process` method but we would have probably - depending on the frequency of incoming 
records - called it way more often, and this could have potentially introduce some performance 
degradation. 
```

**Review by arl — APPROVED** [review]
Good job.
Not much to say on the content this time, just some reformulation, punctuation and grammar.
I accept this so you can directly merge if you address the comments, however you can request another -hopefully final :-) - review if you not only change the form.
